go.mod: upgrade to yaml v4

RaduBerinde opened the pull request:
The https://github.com/go-yaml/yaml (import path `gopkg.in/yaml.v3`)
repository has been archived. The YAML team took over maintenance in a
separate fork at a very confusingly named
https://github.com/yaml/go-yaml (import path `go.yaml.in/yaml/v4`).

This commit updates uses of v3 to `v4.0.0-rc.3` from the new
repository. No code changes are required.

Upgrading v2 uses to v4 will be done separately, as that requires code
changes.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/157800)
<!-- Reviewable:end -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> @sumeerbhola reviewed 50 of 50 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157800#-:-OeITPsWAvZe9G0HKJXK:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @arjunmahishi, @DarrylWong, @golgeek, and @jasonlmfong)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde committed:
go.mod: upgrade to yaml v4

The https://github.com/go-yaml/yaml (import path `gopkg.in/yaml.v3`)
repository has been archived. The YAML team took over maintenance in a
separate fork at a very confusingly named
https://github.com/yaml/go-yaml (import path `go.yaml.in/yaml/v4`).

This commit updates uses of v3 to `v4.0.0-rc.3` from the new
repository. No code changes are required.

Upgrading v2 uses to v4 will be done separately, as that requires code
changes.

Epic: none
Release note: None
